### ISSUE TITLE
[Feature]: Define Echart ToolTip Functions In Javascript Files Instead of Inside Chart Widget

### ISSUE THREAD

**Issue by cch264** [appsmithorg/appsmith#34474]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

I found the thread requesting the ability to pass a function to the echarts option object for the formatter function. I was able to get this working but only by defining an anonymous function in the Custom Echarts Configuration input in the chart widget. This is not ideal because the tooltip function I need to implement is quite large and putting it in that little text box reduces reusability because it has to be defined on demand and reduces readability.

Example:
I have a javascript object called GetAuthObj and a tooltip function called getToolTip defined in this object. When I define my echart configuration I set the formatter option to GetAuthObj.getToolTip but appsmith does not pass the tooltip function correctly. It passes a function that takes no args and causes echarts to error when trying to show the tooltip.

See the image for my config object and the output when running the echart:
![image](https://github.com/appsmithorg/appsmith/assets/77506532/2ef9722c-7ece-443b-8f35-a0dab6e85169)
![image](https://github.com/appsmithorg/appsmith/assets/77506532/0f7e0bc5-0ef4-4271-9139-e0693700ad33)

And the get tooltip function:
![image](https://github.com/appsmithorg/appsmith/assets/77506532/2c4b4a1f-856a-43de-a6c1-ec621e378a34)

This however works:
![image](https://github.com/appsmithorg/appsmith/assets/77506532/8fc452bd-4e9a-4fa1-9900-5871c735997f)

I am running the appsmith application in an iframe if that makes a difference.

### Why should this be worked on?

This promotes re-usability and readability for echart functionality.

**Comment by Nikhil-Nandagopal** [comment]
@cch264 can I suggest simply returning the entire chart configuration as a function and that way you can control any other functions within it that you need to call?

**Comment by cch264** [comment]
Yes that is an option, previously I was passing the chart configuration as an object to the echart widget. But this does not work with formatter functions I get the same error as before. See below. I am defining the chart option as an empty object while waiting for auth credentials to load. Then I set the a property on the javascript object and then bind that property to the widget.

![image](https://github.com/appsmithorg/appsmith/assets/77506532/c119288a-8c17-48a6-aacc-48b77b4bf2bb)
![image](https://github.com/appsmithorg/appsmith/assets/77506532/09ee22f3-e574-4965-be5a-ad904ebe243a)
![image](https://github.com/appsmithorg/appsmith/assets/77506532/fb824f82-c861-4c62-b298-224652879460)

Trying out a function that returns an chart config object same issue:

![image](https://github.com/appsmithorg/appsmith/assets/77506532/16705b52-ca4c-4faa-b691-4dfe59433c83)
![image](https://github.com/appsmithorg/appsmith/assets/77506532/d05704ff-8317-4dff-8ab8-35cc389094e5)
This function returns an options object and the chart loads without issue but the tooltip functionality is still broken
See the return code
`// Specify the configuration items and data for the chart
		var option = {
			legend: {
				top: 50,
				selectedMode: false,
			},
			tooltip: {
				trigger: 'axis',
				axisPointer: {
					type: 'cross',
				},
				formatter: this.getToolTip
			},
			grid: {
				left: 100,
				right: 100,
				top: 100,
				bottom: 200
			},
			dataZoom: [
				{
					type: 'inside',
					start: 0,
					end: 100
				}
			],
			yAxis: {
				type: 'value',
				max: 100
			},
			xAxis: {
				type: 'category',
				data: xAxisLabels,
				axisLabel: {
					interval: 0,
					rotate: 70
				}
			},
			series
		};
		return option;`

It appears the issue is using an object property to pass a function to the formatter property because when I switch to using an anonymous function this works fine. I think the issue still stands because using an anonymous function is not the best solution and is very confusing because my code is valid JavaScript but does not work in this context.

![image](https://github.com/appsmithorg/appsmith/assets/77506532/ddeeb3d7-af7b-43b7-8381-15a82bfb9d65)

**Comment by cch264** [comment]
Going to change the feature request to **Define Echart ToolTip Functions Using Named Functions**
